### PR TITLE
SubTask

### PR DESCRIPTION
**Pull request by pawelknes**
~Added a new field to the `Task` class named `defaults` which sets default values of an instance for chosen fields in both `inputs` and `outputs`.~

Added a new class `SubTask` which allows creating a new task in order to inherit attributes (`metrics`, `inputs` etc.) from other tasks, and possibly add new fields or overwrite others. 
`prediction_type` must be consistent across all combined tasks. Each following defined sub-task has priority over previous ones since its fields may overwrite values for the same keys defined in previous tasks, while any optional parameter defined for the `SubTask` itself has the top priority.

For #893

### CODE REVIEW

File: src/unitxt/task.py [hunk 1]
```diff
@@ -99,9 +105,46 @@ def check_metrics_type(self) -> None:
...
+                )
+
+                assert isoftype(default_value, parse_type_string(val_type)), (
+                    f"The value of '{default_name}' from the 'defaults' must be of "
+                    f"type '{val_type}', however, it is of type '{type(default_value)}'."
+                )
+
+    def set_default_values(self, instance: Dict[str, Any]) -> Dict[str, Any]:
+        if self.defaults:
+            instance.update(self.defaults)
```
Comment (yoavkatz) [inline src/unitxt/task.py:140]: What happens if the value is already set in the instance? In this case, we don't want to value to be updated.

Comment (pawelknes) [reply src/unitxt/task.py:140]: Alright, I thought it should be overwritten anyway. So I'll change that

File: src/unitxt/task.py [hunk 2]
```diff
@@ -119,3 +162,94 @@ def process(
 
 class FormTask(Task):
     pass
+
+
+class SubTask(Task):
```
Comment (yoavkatz) [inline src/unitxt/task.py:167]: Personally, I thinking this kind of multiple inheritance is to complex for users to understand.  Unitxt is already too complicated for most users.   Even in python it's complex, and people will not be able to debug this.   
We are getting a lot of feedback that unitxt becomes too complicated.

I suggest splitting between these changes, the default one is pretty straight forward - and will save people effort (e.g. not specifying text_type in classification, if they don't need to use it in their templates).  

 I think we should discuss the SubTask again - as we already have methods to modify an existing instance using task.qa[metrics=[mymetric]].

**Comment by elronbandel** [comment]
I think it would be easier to comprehend if instead of having SubTask,  every task would have optional `parent_task`. When `parent_task` is assigned all the fields are populated by the task and its parent. So the inputs would be the inputs of the parent task and in addition the inputs of the current task. Similarly for the the defaults field. That way for instance you could have:
```python
add_to_catalog(Task(
    inputs= { "text" : "str", "target_language" : "str"},
    outputs= {"translation" : "str"},
    prediction_type="str",
    metrics=["metrics.bleu"],
), "tasks.translation")

add_to_catalog(Task(
    parent_task="tasks.translation",
    inputs= { "source_language" : "str"},
), "tasks.translation.specific_language")

add_to_catalog(Task(
    parent_task="tasks.translation.specific_language",
    defaults= { "source_language" : "English"},
), "tasks.translation.from_english")
```
I think its simpler mechanism which is easier to understand and can be useful. what do you think @yoavkatz @pawelknes ?

**Comment by yoavkatz** [comment]
We discussed with Elron, and we think the parent task complicates the API, so for now we will not add it.

**Comment by pawelknes** [comment]
> We discussed with Elron, and we think the parent task complicates the API, so for now we will not add it.

Alright

**Comment by yoavkatz** [comment]
> > We discussed with Elron, and we think the parent task complicates the API, so for now we will not add it.
> 
> Alright

So we are closing this for now.

**Pull request closed without merging**